airflow/models/taskinstance.py

```diff
@@ -1345,11 +1354,12 @@ def handle_failure(self, error, test_mode=None, context=None, force_fail=False,
         # Log failure duration
         session.add(TaskFail(task, self.execution_date, self.start_date, self.end_date))
 
-        if context is not None:
-            context['exception'] = error
```

Comment:
this is the biggest semantic change this PR introduces, failure callback won't have access to `context['exception']` anymore because it's not called from a separate task manager process. It is unavoidable because if we don't execute callback from external monitor process, then we won't be able to catch task failures caused by out of memory or segfault.

I think this is a reasonable tradeoff because some of the success/failure callbacks were already invoked from external manager process when states are updated externally, therefore, the contract for accessing `context['exception']` is already partially broken today.

Reply:
Sorry, this makes it a -1 from me. This information is vital when it's available and used by a number of people to send exception messages in Slack hooks. Without this it's much less useful

We need to keep this when it is available, and either document that it's not available in all cases, or add a "synthetic" exception when it's missing (`AirflowError("state changed to failed")` for example)

Reply:
We can remove this now can't we?

Reply:
yes, `context['exception']` is not set in `_run_finished_callback`.